src/Init/Data/String/Basic.lean

```diff
@@ -290,73 +290,111 @@ where go (acc : String) (s : String) : List String → String
   | a :: as => go (acc ++ s ++ a) s as
   | []      => acc
 
-/-- Iterator for `String`. That is, a `String` and a position in that string. -/
+/-- Iterator over the characters (`Char`) of a `String`.
+
+Typically created by `s.iter`, where `s` is a `String`. -/
 structure Iterator where
+  /-- The string the iterator is for. -/
   s : String
+  /-- The current position.
+
+  This position is not necessarily legal for the string, for instance if one keeps calling
+  `Iterator.next` when `Iterator.atEnd` is true. If the position is not a legal one, then the
+  current character is `(default : Char)`, similar to `String.get` on an illegal position. -/
   i : Pos
   deriving DecidableEq
 
+/-- The iterator positioned at the first character of a string (if any). -/
 def mkIterator (s : String) : Iterator :=
   ⟨s, 0⟩
 
+@[inherit_doc mkIterator]
 abbrev iter := mkIterator
 
+/-- The notion of size for a string iterator is the number of bytes remaining. -/
 instance : SizeOf String.Iterator where
   sizeOf i := i.1.utf8ByteSize - i.2.byteIdx
 
 theorem Iterator.sizeOf_eq (i : String.Iterator) : sizeOf i = i.1.utf8ByteSize - i.2.byteIdx :=
   rfl
 
 namespace Iterator
+/-- The *full* string the iterator is for. -/
 def toString : Iterator → String
   | ⟨s, _⟩ => s
 
+/-- Number of bytes remaining in the iterator. -/
 def remainingBytes : Iterator → Nat
   | ⟨s, i⟩ => s.endPos.byteIdx - i.byteIdx
 
+/-- Current character position. -/
 def pos : Iterator → Pos
   | ⟨_, i⟩ => i
 
+/-- The character at the current position.
+
+On an illegal position, returns `(default : Char)`. -/
 def curr : Iterator → Char
   | ⟨s, i⟩ => get s i
 
+/-- Increases the iterator's position by one, unconditionally. -/
 def next : Iterator → Iterator
   | ⟨s, i⟩ => ⟨s, s.next i⟩
 
+/-- Decreases the iterator's position.
+
+If the position is zero, this function is the identity. -/
 def prev : Iterator → Iterator
   | ⟨s, i⟩ => ⟨s, s.prev i⟩
 
+/-- True if the iterator is past the string's last character. -/
 def atEnd : Iterator → Bool
   | ⟨s, i⟩ => i.byteIdx ≥ s.endPos.byteIdx
 
+/-- True if the iterator is not past the string's last character. -/
 def hasNext : Iterator → Bool
   | ⟨s, i⟩ => i.byteIdx < s.endPos.byteIdx
 
+/-- True if the position is not zero. -/
 def hasPrev : Iterator → Bool
   | ⟨_, i⟩ => i.byteIdx > 0
 
+/-- Replaces the current character in the string.
```

Comment:
```suggestion
/-- Construct an iterator into a string in which the current character is replaced with some other character.

If the iterator contains the only reference to the string, then this will mutate the string rather than allocating a new one.
```

Reply:
Are you sure about the "Constructs an iterator..." description?

I find it verbose and daunting as a first description, IMHO it hides the human-level abstraction of the function which is the current description.

I'm also worried it can be confusing because it is inconsistent with other iterator-transforming functions such as `Iterator.next`, which does not really "move the position", but rather (in general) "constructs an iterator ...".

Reply:
A reasonable compromise could be to have your "Constructs an iterator ..." below the current description, as a more detailed explanation.

I still think that's a bit inconsistent if, for instance, `Iterator.next` does not have a similar clarification, though I think most readers at this point should have at least an intuitive understanding of what "purely functional" means and don't need the clarification anyways.

Reply:
I think it's fine to not say "contructs" here.

I think the bit about mutation is important to have, but the original first paragraph is fine.

Reply:
Okay, the discussion on mutation (assuming that's the "immutable beans" bit) is already in there, slightly reworded.